### PR TITLE
Hardhat deploy

### PR DESCRIPTION
**Pull request by PowVT**
New Features:
- Generate Externally owned EVM account
- View wallet address and balances across all hardhat networks
- implement `hardhat-deploy` plugin by [wighawag](https://github.com/wighawag/hardhat-deploy)

Notes:
- Hardhat-deploy does not currently support UUPS proxies, only transparent. Thus, cannot use the hardhat-deploy `deploy` method for contracts which use UUPS upgradability and need to use the deployProxy function from hardhat 'upgrades'. A main drawback here is the contract address and ABI do not get stored in the deployments folder like they would if the deployed with the plugin. 
> - The large 'protocol' deploy script is a symptom of this and needs to be done this was so it can pass around the contract addresses that would not get stored in the deployments folder. 
> - The `proxy` tag specified in wigahwag's documentation is supposed to call the initializer in a transparent proxy. The error is due to the plugin proxy contract having a function with the same function as one of the OZ hardhat upgrade functions. `Error: function "upgradeTo" will shadow "upgradeTo". Please update code to avoid conflict.`
- Additional documentation in the docs folder.

### CODE REVIEW

File: deploy/01_deploy_vault_factory.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  const { deployer } = await getNamedAccounts();
+
+  /// ===== VaultTemplate =====
+  const vaultTemplate = await deploy('AssetVault', {
+    from: deployer,
+    log: true,
+  });
+  console.log("Using vaultTemplate deployed at: ", vaultTemplate.address)
+
+  const whitelist = await ethers.getContract("CallWhitelist", deployer);
```
Comment (kkennis) [inline deploy/01_deploy_vault_factory.ts:18]: It kind of feels like `CallWhitelist` should just be deployed in this file

File: deploy/00_deploy_protocol.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  /// ===== RepaymentController =====
+  const repaymentController = await deploy('RepaymentController', {
+    from: deployer,
+    args: [loanCore.address, borrowerNote.address, lenderNote.address],
+    log: true,
+  });
+
+
+  /// ===== GrantRoles =====
```
Comment (kkennis) [inline deploy/00_deploy_protocol.ts:69]: There are also roles to grant on:

1) The borrower and lender note (by calling `initialize`)
2) Any upgradeable contract needs the correct owner

Take a look at Mouzayan's open PR for all the function calls we need for correct ownership and role setup. Maybe it makes sense to kind of combine these PRs and just change her script's filename and make it numbered (e.g. `04_transfer_ownership.ts`)?

File: hardhat.config.ts [hunk 3]
```diff
@@ -128,6 +133,9 @@ const config: HardhatUserConfig = {
         artifacts: "./artifacts",
         cache: "./cache",
         sources: "./contracts",
+        deployments: './deployments',
```
Comment (kkennis) [inline hardhat.config.ts:136]: When I pull down the branch, and try to run a task (e.g. `yarn hardhat accounts`), I get an error from this line:

```
➜  pawnfi-contracts git:(hardhat-deploy) yarn hardhat accounts
yarn run v1.22.17
$ /Users/kevin/dev/pawnfi-contracts/node_modules/.bin/hardhat accounts
An unexpected error occurred:

hardhat.config.ts:136:9 - error TS2322: Type '{ artifacts: string; cache: string; sources: string; deployments: string; deploy: string; imports: string; tests: string; }' is not assignable to type 'ProjectPathsUserConfig'.
  Object literal may only specify known properties, and 'deployments' does not exist in type 'ProjectPathsUserConfig'.

136         deployments: './deployments',
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/hardhat/types/config.d.ts:168:5
    168     paths?: ProjectPathsUserConfig;
            ~~~~~
    The expected type comes from property 'paths' which is declared here on type 'HardhatUserConfig'

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: tasks/functions/generate.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  const hdwallet = hdkey.fromMasterSeed(seed);
+  const walletHdPath = "m/44'/60'/0'/0/";
+  const accountIndex = 0;
+  const fullPath = walletHdPath + accountIndex.toString();
+  debugLog('FULL PATH', fullPath);
+  const wallet = hdwallet.derivePath(fullPath).getWallet();
+  const privateKey = `0x${wallet.getPrivateKey().toString('hex')}`;
+  debugLog('PRIVATE KEY', privateKey);
+  const address = `0x${privateToAddress(wallet.getPrivateKey()).toString('hex')}`;
+  console.log(`🔐 Account Generated as ${address} and set as mnemonic in packages/hardhat`);
```
Comment (kkennis) [inline tasks/functions/generate.ts:27]: Curious - what's the source of these task files? Do they come from another repo/template somewhere?

**Review by kkennis — COMMENTED** [review]
Nice to see how the `deploy` folder works, but I'd like to try it locally: see the comment about me being unable to build.

File: hardhat.config.ts (hunk 3 above)
Comment (kkennis) [reply hardhat.config.ts:136]: Fixed after running `yarn install`

File: deploy/01_deploy_vault_factory.ts (hunk 1 above)
Comment (PowVT) [reply deploy/01_deploy_vault_factory.ts:18]: Yup, I agree here. Will move it. Thanks

**Pull request closed without merging**